File: src/mpu/lib/miro.py
"""Тонкий клиент Miro REST API v2.

Используется только из mpu-d2-miro. Намеренно без новых зависимостей —
urllib из stdlib + json. С rate-limit retry на 429.
"""

from __future__ import annotations

import json
import sys
import time
from dataclasses import dataclass
from typing import Any
from urllib.error import HTTPError
from urllib.parse import quote
from urllib.request import Request, urlopen

API_BASE = "https://api.miro.com"


@dataclass
class FrameRef:
    id: str
    title: str
    x: float  # center
    y: float
    w: float
    h: float


class MiroClient:
    def __init__(self, token: str, board_id_raw: str):
        self.token = token
        self.board_id = quote(board_id_raw, safe="")
        self.base = f"{API_BASE}/v2/boards/{self.board_id}"

    def _request(
        self, method: str, path: str, body: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        url = f"{self.base}{path}" if path.startswith("/") else f"{API_BASE}{path}"
        data = json.dumps(body).encode("utf-8") if body is not None else None
        headers = {
            "Authorization": f"Bearer {self.token}",
            "Accept": "application/json",
        }
        if data is not None:
            headers["Content-Type"] = "application/json"

        backoff = 1.0
        for _ in range(6):
            req = Request(url, data=data, method=method, headers=headers)
            try:
                with urlopen(req) as r:
                    txt = r.read().decode("utf-8")
                    return json.loads(txt) if txt else {}
            except HTTPError as e:
                err_body = e.read().decode("utf-8", "replace")
                if e.code == 429:
                    wait = int(e.headers.get("Retry-After", str(int(backoff))))
                    print(f"[miro] 429 rate-limit, sleep {wait}s", file=sys.stderr)
                    time.sleep(wait)
                    backoff = min(backoff * 2, 30)
                    continue
                raise MiroAPIError(method, path, e.code, err_body) from None
        raise MiroAPIError(method, path, 429, "exhausted retries")

    # ---------- frames ----------

    def list_frames(self) -> list[FrameRef]:
        out: list[FrameRef] = []
        cursor: str | None = None
        while True:
            qs = "?limit=50" + (f"&cursor={cursor}" if cursor else "")
            res = self._request("GET", f"/items{qs}&type=frame")
            for it in res.get("data", []):
                if it.get("type") != "frame":
                    continue
                pos = it.get("position", {}) or {}
                geo = it.get("geometry", {}) or {}
                title = ((it.get("data") or {}).get("title")) or ""
                out.append(
                    FrameRef(
                        id=it["id"],
                        title=title,
                        x=float(pos.get("x", 0)),
                        y=float(pos.get("y", 0)),
                        w=float(geo.get("width", 0)),
                        h=float(geo.get("height", 0)),
                    )
                )
            cursor = res.get("cursor")
            if not cursor:
                break
        return out

    def find_frame_by_title(self, title: str) -> FrameRef | None:
        for f in self.list_frames():
            if f.title == title:
                return f
        return None

    def rightmost_frame_edge(self) -> tuple[float, float]:
        """Возвращает (max_right_x, средний_y) всех существующих фреймов.
        Если фреймов нет, возвращает (0, 0)."""
        frames = self.list_frames()
        if not frames:
            return 0.0, 0.0
        right = max(f.x + f.w / 2 for f in frames)
        avg_y = sum(f.y for f in frames) / len(frames)
        return right, avg_y

    def unlock_item(self, item_id: str) -> None:
        """Снять lock с item'а через PATCH. На lock-able элементах работает безопасно;
        на уже-разлоченных тоже не бросает (Miro возвращает 200)."""
        try:
            self._request("PATCH", f"/items/{item_id}", {"locked": False})
        except MiroAPIError as e:
            # 404 — item уже удалён; 400 — на некоторых типах PATCH не поддерживается
            if e.status not in (400, 404):
                raise

    def delete_frame(self, frame_id: str) -> None:
        """Удаляет фрейм со всем содержимым.

        Miro не удаляет children при `DELETE /frames/{id}` — они остаются на доске
        как orphan-элементы. Поэтому сначала собираем всех children через
        /items?parent_item_id=... и удаляем их, потом сам фрейм.

        Залоченные элементы и сам залоченный фрейм автоматически разлочиваются перед
        удалением (lock в Miro REST API виден как поле items.locked, переключается
        через PATCH /items/{id}).
        """
        for item_id, _item_type in self._iter_children(frame_id):
            try:
                self._request("DELETE", f"/items/{item_id}")
            except MiroAPIError as e:
                if e.status == 404:
                    continue
                if e.status == 400 and "locked" in e.body.lower():
                    self.unlock_item(item_id)
                    try:
                        self._request("DELETE", f"/items/{item_id}")
                    except MiroAPIError as e2:
                        if e2.status != 404:
                            raise
                    continue
                raise
        try:
            self._request("DELETE", f"/frames/{frame_id}")
        except MiroAPIError as e:
            if e.status == 404:
                return
            if e.status == 400 and "locked" in e.body.lower():
                self.unlock_item(frame_id)
                try:
                    self._request("DELETE", f"/frames/{frame_id}")
                except MiroAPIError as e2:
                    if e2.status == 404:
                        return
                    raise
                return
            raise

    def _iter_children(self, frame_id: str) -> list[tuple[str, str]]:
        """Возвращает список (item_id, item_type) для всех элементов внутри фрейма."""
        out: list[tuple[str, str]] = []
        cursor: str | None = None
        while True:
            qs = f"?parent_item_id={frame_id}&limit=50"
            if cursor:
                qs += f"&cursor={cursor}"
            res = self._request("GET", f"/items{qs}")
            for it in res.get("data", []) or []:
                out.append((it["id"], it.get("type", "")))
            cursor = (res.get("cursor") or "") or None
            if not cursor:
                break
        return out

    @staticmethod
    def _delete_endpoint_for(item_type: str) -> str:
        # Большинство типов — общий /items/{id}, но shape/connector/text/sticky_note и
        # frame имеют типизированные эндпоинты. /items/{id} безопаснее.
        return "/items"

    def create_frame(
        self, *, title: str, x: float, y: float, width: float, height: float
    ) -> FrameRef:
        body = {
            "data": {"title": title, "format": "custom", "type": "freeform"},
            "style": {"fillColor": "#ffffff"},
            "position": {"x": x, "y": y},
            "geometry": {"width": width, "height": height},
        }
        res = self._request("POST", "/frames", body)
        return FrameRef(id=res["id"], title=title, x=x, y=y, w=width, h=height)

    # ---------- shapes / connectors / texts ----------

    def create_shape(
        self,
        *,
        parent_id: str,
        kind: str,
        content_html: str,
        x: float,
        y: float,
        width: float,
        height: float,
        fill: str = "#ffffff",
        fill_opacity: str = "1.0",
        border_color: str = "#1a1a1a",
        border_width: str = "1",
        border_style: str = "normal",
        font_size: str = "12",
        text_align: str = "center",
        text_align_vertical: str = "middle",
    ) -> str:
        body = {
            "data": {"shape": kind, "content": content_html},
            "style": {
                "fillColor": fill,
                "fillOpacity": fill_opacity,
                "borderColor": border_color,
                "borderWidth": border_width,
                "borderStyle": border_style,
                "color": "#1a1a1a",
                "fontSize": font_size,
                "textAlign": text_align,
                "textAlignVertical": text_align_vertical,
            },
            "position": {"x": x, "y": y},
            "geometry": {"width": max(width, 60), "height": max(height, 40)},
            "parent": {"id": parent_id},
        }
        res = self._request("POST", "/shapes", body)
        return res["id"]

    def create_card(
        self,
        *,
        parent_id: str,
        title: str,
        description: str,
        x: float,
        y: float,
        width: float,
        height: float,
        theme: str = "#2d9bf0",
    ) -> str:
        """Создать Miro card. Description поддерживает ограниченный HTML
        (p/strong/em/ul/li/br). theme = hex цвет «полоски» слева у card."""
        # Miro REST API требует width >= 256 для cards (валидация на сервере).
        body: dict[str, Any] = {
            "data": {"title": title, "description": description},
            "style": {"cardTheme": theme},
            "position": {"x": x, "y": y},
            "geometry": {"width": max(width, 256), "height": max(height, 40)},
            "parent": {"id": parent_id},
        }
        res = self._request("POST", "/cards", body)
        return res["id"]

    def create_text(
        self,
        *,
        parent_id: str,
        content_html: str,
        x: float,
        y: float,
        width: float,
    ) -> str:
        body = {
            "data": {"content": content_html},
            "position": {"x": x, "y": y},
            "geometry": {"width": max(width, 200)},
            "parent": {"id": parent_id},
        }
        res = self._request("POST", "/texts", body)
        return res["id"]

    def create_connector(
        self,
        *,
        src_id: str,
        dst_id: str,
        label: str = "",
        shape: str = "elbowed",
        snap_start: str | None = None,
        snap_end: str | None = None,
    ) -> str:
        """Создать connector. snap_start/snap_end ∈ {top, bottom, left, right, auto}.

        Полезно при bidirectional парах (A→B и B→A): одной паре указать `top`,
        другой — `bottom`, чтобы линии не накладывались.
        """
        start_item: dict[str, Any] = {"id": src_id}
        end_item: dict[str, Any] = {"id": dst_id}
        if snap_start:
            start_item["snapTo"] = snap_start
        if snap_end:
            end_item["snapTo"] = snap_end
        body: dict[str, Any] = {
            "startItem": start_item,
            "endItem": end_item,
            "shape": shape,
            "style": {"strokeColor": "#1a1a1a", "strokeWidth": "1"},
        }
        if label:
            body["captions"] = [{"content": label}]
        res = self._request("POST", "/connectors", body)
        return res["id"]


class MiroAPIError(Exception):
    def __init__(self, method: str, path: str, status: int, body: str):
        self.method = method
        self.path = path
        self.status = status
        self.body = body
        super().__init__(f"miro {method} {path} -> {status}: {body[:300]}")

File: ts/back/src/d2miro/miro.ts
/**
 * Клиент Miro API — граница, снятая с живой службы
 * (`docs/specs/d2-miro.md`, «Снято с живой службы 2026-08-31»;
 * фикстуры — `fixtures/d2-miro/live/`).
 *
 * Отпечатки службы, ради которых клиент выглядит именно так:
 *
 * - `POST /frames|/shapes|/texts` отвечает `201`, а `POST /connectors`
 *   — `200`. Поэтому успех определяется диапазоном `2xx`, а не
 *   списком кодов: список пришлось бы держать в двух местах.
 * - `POST /frames` отвечает `500` примерно в половине попыток при
 *   живой квоте (`x-ratelimit-remaining` 97 600 из 100 000), и
 *   повтор лечит. Отсюда главное расхождение с оригиналом: клиент
 *   повторяет **и `5xx`**, а не только `429` (оригинал —
 *   `py/src/mpu/lib/http_retry.py:36`), и **считает повторы**, чтобы
 *   итог назвал их числом.
 * - `DELETE /frames/{id}` детей не удаляет: они остаются сиротами с
 *   `parent: null`. Обход детей — обязанность вызывающего
 *   (`render.ts`), клиент этого за него не делает.
 * - Повторное удаление отвечает `404` с кодом `"3.0201"`; для
 *   удаления это успех, а не отказ.
 *
 * Токен живёт только в заголовке: в тексты ошибок идут метод, путь и
 * тело ответа — инвариант спеки «токен не попадает ни в вывод, ни в
 * тексты ошибок».
 */

/** Отказ службы, который повтором не лечится. */
export class MiroError extends Error {
  override readonly name = "MiroError";
  constructor(readonly status: number, message: string) {
    super(message);
  }
}

/** Минимум от `fetch`, который нужен клиенту. */
export type FetchLike = (
  url: string,
  init: { method: string; headers: Record<string, string>; body?: string },
) => Promise<Response>;

/** Зависимости клиента: сеть, сон и диагностическая строка. */
export interface MiroIo {
  readonly fetch: FetchLike;
  /** Пауза между попытками; в тестах — без настоящего сна. */
  readonly sleep: (ms: number) => Promise<void>;
  /** Строка `[miro] …` в stderr. */
  readonly note: (line: string) => void;
}

/** Элемент доски в том виде, в каком он нужен рендеру. */
export interface MiroItem {
  readonly id: string;
  readonly type: string;
  readonly title?: string;
  readonly x?: number;
  readonly y?: number;
  readonly width?: number;
  readonly height?: number;
}

/** Потолок попыток одного запроса (спека: максимум 6). */
const MAX_ATTEMPTS = 6;
/** Потолок паузы между попытками, секунды. */
const MAX_SLEEP_S = 30;
/** Длина тела ответа в тексте ошибки (спека). */
const BODY_LIMIT = 300;
/** Размер страницы листинга; курсор — в ответе. */
const PAGE = 50;

/** Пауза перед попыткой номер `attempt` (1-я повторная — 1 с). */
function backoffSeconds(attempt: number): number {
  return Math.min(2 ** (attempt - 1), MAX_SLEEP_S);
}

/** Тело ответа для текста ошибки — обрезанное, без заголовков. */
function shortBody(text: string): string {
  return text.length <= BODY_LIMIT ? text : text.slice(0, BODY_LIMIT);
}

/** Клиент одной доски. */
export class MiroBoard {
  /** Сколько раз запрос пришлось повторить за жизнь клиента. */
  #retries = 0;
  readonly #base: string;

  constructor(
    private readonly io: MiroIo,
    boardId: string,
    private readonly token: string,
  ) {
    this.#base = `https://api.miro.com/v2/boards/${
      encodeURIComponent(boardId)
    }`;
  }

  /** Число повторов — итог обязан называть его числом (спека). */
  get retries(): number {
    return this.#retries;
  }

  /**
   * Запрос с повтором на `429` и `5xx`. Возвращает тело и статус —
   * решение «успех или отказ» принимает вызывающий метод: у удаления
   * `404` успех, у прочих нет.
   */
  async request(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<{ status: number; text: string }> {
    let lastNonRetryable: { status: number; text: string } | undefined;
    for (let attempt = 1; attempt <= MAX_ATTEMPTS; attempt++) {
      const response = await this.send(method, path, body);
      const text = await response.text();
      if (!retryable(response.status)) {
        lastNonRetryable = { status: response.status, text };
        break;
      }
      if (attempt === MAX_ATTEMPTS) {
        lastNonRetryable = { status: response.status, text };
        break;
      }
      await this.pause(response, attempt);
    }
    if (lastNonRetryable === undefined) {
      // Недостижимо: цикл выходит только через присваивание. Ветка
      // оставлена вместо `!`, чтобы отсутствие ответа не превращалось
      // в молчаливый `undefined` у вызывающего.
      throw new MiroError(0, `miro ${method} ${path} -> нет ответа`);
    }
    return lastNonRetryable;
  }

  /**
   * Один запрос к службе. Сбой самого обращения — не ответ, а отказ
   * своим классом: без этого он уходил бы наружу «unexpected error» с
   * трейсом, чего спека запрещает (отклонение-fix). Повтора у него
   * нет: сеть повторит вызывающий, а неверный заголовок повторять
   * бессмысленно — замер 2026-08-31, токен с кириллицей роняет
   * `fetch` до всякой сети.
   */
  private async send(
    method: string,
    path: string,
    body: unknown,
  ): Promise<Response> {
    try {
      return await this.io.fetch(`${this.#base}${path}`, {
        method,
        headers: {
          "authorization": `Bearer ${this.token}`,
          "accept": "application/json",
          ...(body === undefined ? {} : { "content-type": "application/json" }),
        },
        ...(body === undefined ? {} : { body: JSON.stringify(body) }),
      });
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      throw new MiroError(
        0,
        `miro ${method} ${path}: transport error: ${message}`,
      );
    }
  }

  /** Пауза перед повтором с диагностической строкой. */
  private async pause(response: Response, attempt: number): Promise<void> {
    this.#retries++;
    const header = Number(response.headers.get("retry-after"));
    const seconds = response.status === 429 && Number.isFinite(header) &&
        header > 0
      ? Math.min(header, MAX_SLEEP_S)
      : backoffSeconds(attempt);
    this.io.note(
      response.status === 429
        ? `[miro] 429 rate-limit, sleep ${seconds}s`
        : `[miro] ${response.status} from service, retry ${attempt} in ` +
          `${seconds}s`,
    );
    await this.io.sleep(seconds * 1000);
  }

  /** Запрос, у которого не-2xx — отказ; отдаёт разобранный JSON. */
  async json<T>(method: string, path: string, body?: unknown): Promise<T> {
    const { status, text } = await this.request(method, path, body);
    if (status < 200 || status >= 300) {
      throw new MiroError(
        status,
        `miro ${method} ${path} -> ${status}: ${shortBody(text)}`,
      );
    }
    return (text === "" ? {} : JSON.parse(text)) as T;
  }

  /** Все фреймы доски — полной пагинацией, а не первой страницей. */
  async frames(): Promise<readonly MiroItem[]> {
    return await this.page("/items?type=frame");
  }

  /** Дети фрейма — тоже полной пагинацией. */
  async children(frameId: string): Promise<readonly MiroItem[]> {
    return await this.page(
      `/items?parent_item_id=${encodeURIComponent(frameId)}`,
    );
  }

  /**
   * Обход страниц по курсору. Страница обрезает молча — сравнение
   * первых `limit` элементов дало бы ложное «состав совпал» (замер
   * спецификатора 2026-08-31).
   */
  private async page(path: string): Promise<readonly MiroItem[]> {
    const items: MiroItem[] = [];
    let cursor: string | undefined;
    do {
      const query = `${path}&limit=${PAGE}` +
        (cursor === undefined ? "" : `&cursor=${encodeURIComponent(cursor)}`);
      const reply = await this.json<
        { data?: unknown[]; cursor?: string }
      >("GET", query);
      for (const raw of reply.data ?? []) items.push(itemOf(raw));
      cursor = reply.cursor === "" ? undefined : reply.cursor;
    } while (cursor !== undefined);
    return items;
  }

  /** Создание элемента: `201` у фрейма и шейпа, `200` у коннектора. */
  async create(path: string, body: unknown): Promise<string> {
    const created = await this.json<{ id?: string }>("POST", path, body);
    if (typeof created.id !== "string" || created.id === "") {
      throw new MiroError(0, `miro POST ${path}: в ответе нет id`);
    }
    return created.id;
  }

  /** Снятие блокировки; `400` и `404` терпимы (спека). */
  async unlock(id: string): Promise<void> {
    const { status } = await this.request(
      "PATCH",
      `/items/${encodeURIComponent(id)}`,
      { data: { locked: false } },
    );
    if (status === 400 || status === 404 || (status >= 200 && status < 300)) {
      return;
    }
    throw new MiroError(status, `miro PATCH /items/${id} -> ${status}`);
  }

  /**
   * Удаление элемента или фрейма: `404` — успех (уже удалён), `400` со
   * словом `locked` — разлочить и повторить один раз.
   */
  async remove(path: string, id: string): Promise<void> {
    const target = `${path}/${encodeURIComponent(id)}`;
    const first = await this.request("DELETE", target);
    if (deleted(first.status)) return;
    if (first.status === 400 && first.text.includes("locked")) {
      await this.unlock(id);
      const second = await this.request("DELETE", target);
      if (deleted(second.status)) return;
      throw new MiroError(
        second.status,
        `miro DELETE ${target} -> ${second.status}: ${shortBody(second.text)}`,
      );
    }
    throw new MiroError(
      first.status,
      `miro DELETE ${target} -> ${first.status}: ${shortBody(first.text)}`,
    );
  }
}

/** Повторяемый исход: троттлинг и любой отказ самой службы. */
function retryable(status: number): boolean {
  return status === 429 || status >= 500;
}

/** Удаление удалось: 2xx либо «уже удалён». */
function deleted(status: number): boolean {
  return (status >= 200 && status < 300) || status === 404;
}

/** Элемент ответа службы в том виде, в каком он нужен рендеру. */
function itemOf(raw: unknown): MiroItem {
  const item = raw as {
    id?: unknown;
    type?: unknown;
    data?: { title?: unknown };
    position?: { x?: unknown; y?: unknown };
    geometry?: { width?: unknown; height?: unknown };
  };
  return {
    id: String(item.id ?? ""),
    type: String(item.type ?? ""),
    title: typeof item.data?.title === "string" ? item.data.title : undefined,
    x: numberOf(item.position?.x),
    y: numberOf(item.position?.y),
    width: numberOf(item.geometry?.width),
    height: numberOf(item.geometry?.height),
  };
}

function numberOf(value: unknown): number | undefined {
  return typeof value === "number" ? value : undefined;
}
